managed/services/telemetry/telemetry_test.go

```diff
@@ -0,0 +1,363 @@
+				start:  now,
+				config: getTestConfig(true, testSourceName, 500*time.Millisecond+2*time.Second),
+				dsRegistry: mockDataSourceLocator(t, [][]*pmmv1.ServerMetric_Metric{expectedServerMetrics},
+					testSourceName, 2),
+				pmmVersion: pmmVersion,
+				dus: func(l *logrus.Entry) distributionUtilService {
+					var dusMock mockDistributionUtilService
+					dusMock.Test(t)
+					dusMock.On("getDistributionMethodAndOS", l).
+						Return(serverpb.DistributionMethod_AMI, pmmv1.DistributionMethod_AMI, "ami").
+						Times(2)
+					t.Cleanup(func() {
+						dusMock.AssertExpectations(t)
+					})
+					return &dusMock
+				},
+			},
+			mockTelemetrySender: initMockTelemetrySender(t, expectedReport, 2),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			ctx, cancel := context.WithTimeout(context.Background(), tt.testTimeout)
+			defer cancel()
+
+			logger := logrus.StandardLogger()
+			logger.SetLevel(logrus.DebugLevel)
+			logEntry := logrus.NewEntry(logger)
+
+			s := Service{
+				db:                  tt.fields.db(),
+				l:                   logEntry,
+				start:               tt.fields.start,
+				config:              tt.fields.config,
+				dsRegistry:          tt.fields.dsRegistry(),
+				pmmVersion:          "",
+				os:                  tt.fields.os,
+				sDistributionMethod: 0,
+				tDistributionMethod: 0,
+				dus:                 tt.fields.dus(logEntry),
+				portalClient:        tt.mockTelemetrySender(),
+			}
+
+			var wg sync.WaitGroup
+
+			wg.Add(1)
+			go func() {
+				s.Run(ctx)
+				wg.Done()
+			}()
+
+			wg.Wait()
+		})
+	}
+}
+
+func initMockTelemetrySender(t *testing.T, expetedReport *reporter.ReportRequest, timesCall int) func() sender {
+	return func() sender {
+		var mockTelemetrySender mockSender
+		mockTelemetrySender.Test(t)
+		mockTelemetrySender.On("SendTelemetry",
+			mock.AnythingOfType(reflect.TypeOf(context.TODO()).Name()),
+			mock.MatchedBy(func(report *reporter.ReportRequest) bool {
+				return equalReports(report, expetedReport)
+			}),
+		).
+			Return(nil).
+			Times(timesCall)
+		t.Cleanup(func() {
+			mockTelemetrySender.AssertExpectations(t)
+		})
+
+		return &mockTelemetrySender
+	}
+}
+
+func equalReports(report *reporter.ReportRequest, expectedReport *reporter.ReportRequest) bool {
+	if len(expectedReport.Metrics) != len(report.Metrics) {
+		return false
+	}
+
+	for i, m := range expectedReport.Metrics {
+		gotMetric := report.Metrics[i]
+
+		if m.PmmServerVersion != gotMetric.PmmServerVersion {
+			return false
+		}
+		if m.DistributionMethod != gotMetric.DistributionMethod {
+			return false
+		}
+		for j, mm := range m.Metrics {
+			if mm.Key != gotMetric.Metrics[j].Key {
+				return false
+			}
+			if mm.Value != gotMetric.Metrics[j].Value {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func mockDataSourceLocator(t *testing.T, metrics [][]*pmmv1.ServerMetric_Metric, testSourceName string, times int) func() DataSourceLocator {
```

Comment:
It looks like to grab some real metrics is easier than mock datasources

Reply:
do you mean to hardcode real metrics in the code? or grab from some real datasource in the unit tests these metrics?

Reply:
Use real datasource.